pkg/leech/pipeline.go

```diff
@@ -0,0 +1,269 @@
+	JSON_VALUE(payload, "$.repository.owner.login") org_name,
+	JSON_VALUE(payload, "$.workflow_run.logs_url") logs_url,
+	JSON_VALUE(payload, "$.workflow_run.actor.login") github_actor,
+	JSON_VALUE(payload, "$.workflow_run.html_url") workflow_url
+FROM ` + "`%s`" + `
+WHERE
+event = "workflow_run"
+AND JSON_VALUE(payload, "$.workflow_run.status") = "completed"
+AND delivery_id NOT IN (
+SELECT
+  delivery_id
+FROM ` + "`%s`" + `
+LIMIT %d)
+LIMIT %d
+`
+
+// errLogsExpired is a marker error so that upstream processing knows
+// that the logs for a given event no longer exist.
+var errLogsExpired = errors.New("GitHub logs expired")
+
+// Pipeline represents an implementation of Beam Pipeline that queries
+// for events, looks for their workflow logs in GitHub and then copies them
+// to Google Cloud Storage.
+type Pipeline struct {
+	cfg     *Config
+	ghApp   *githubapp.GitHubApp
+	storage *ObjectStore
+}
+
+// ingestLogsFn is an object that implements beams "DoFn" interface to
+// provide the main processing of the event.
+type ingestLogsFn struct {
+	pipe *Pipeline
+}
+
+// BeamInit handles preregistration of pipeline shapes with their associated
+// type signature and shape to speed up runtime execution.
+func BeamInit() {
+	register.DoFn1x1[EventRecord, LeechRecord](&ingestLogsFn{})
+}
+
+// New creates a new instance of the leech Pipeline.
+func New(ctx context.Context, cfg *Config) (*Pipeline, error) {
+	store, err := NewObjectStore(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create object store client: %w", err)
+	}
+	pk, err := readPrivateKey(cfg.GitHubPrivateKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read private key: %w", err)
+	}
+	ghAppConfig := githubapp.NewConfig(cfg.GitHubAppID, cfg.GitHubInstallID, pk)
+	ghApp := githubapp.New(ghAppConfig)
+
+	return &Pipeline{
+		cfg:     cfg,
+		ghApp:   ghApp,
+		storage: store,
+	}, nil
+}
+
+// Prepare connects the internal pipeline implementation to the beam
+// scope object associated with the primary beam Pipeline.
+func (pipe *Pipeline) Prepare(scope beam.Scope) {
+	// BigQuery table notation is not consistent so it needs represented in a few different
+	// formats to appease the BigQuery client.
+	eventsTableDotNotation := fmt.Sprintf("%s.%s", pipe.cfg.EventsProjectID, pipe.cfg.EventsTable)
+	leechTableDotNotation := fmt.Sprintf("%s.%s", pipe.cfg.LeechProjectID, pipe.cfg.LeechTable)
+	leechTableColonNotation := fmt.Sprintf("%s:%s", pipe.cfg.LeechProjectID, pipe.cfg.LeechTable)
+	batchSize := pipe.cfg.BatchSize
+
+	var event EventRecord
+	query := fmt.Sprintf(sourceQuery, eventsTableDotNotation, leechTableDotNotation, batchSize, batchSize)
+	// step 1: query BigQuery for unprocessed events
+	col := bigqueryio.Query(scope, pipe.cfg.LeechProjectID, query, reflect.TypeOf(event), bigqueryio.UseStandardSQL())
+	// step 2: process the events in parallel, ingesting logs
+	res := beam.ParDo(scope, &ingestLogsFn{pipe: pipe}, col)
```

Comment:
Curious what kind of "parallel" is this? Fan out to multiple running containers? Fan out to multiple threads?

Reply:
Beam will determine if it needs multiple workers or if it can do it all in 1 worker with multiple threads. We don't have to manage that, the runner makes that choice for us.